dbt/adapters/base/connections.py

```diff
@@ -0,0 +1,289 @@
+
+            if not self.available:
+                raise dbt.exceptions.InternalException(
+                    'Tried to request a new connection "{}" but '
+                    'the maximum number of connections are already '
+                    'allocated!'.format(name)
+                )
+
+            connection = self.available.pop()
+            # connection is temporarily neither in use nor available, but both
+            # collections are in a sane state, so we can release the lock.
+
+        # this potentially calls open(), but does so without holding the lock
+        connection = self.assign(connection, name)
+
+        with self.lock:
+            if name in self.in_use:
+                raise dbt.exceptions.InternalException(
+                    'Two threads concurrently tried to get the same name: {}'
+                    .format(name)
+                )
+            self.in_use[name] = connection
+
+        return connection
+
+    @abc.abstractmethod
+    def cancel_open(self):
+        """Cancel all open connections on the adapter. (passable)"""
+        raise dbt.exceptions.NotImplementedException(
+            '`cancel_open` is not implemented for this adapter!'
+        )
+
+    @abstractclassmethod
+    def open(cls, connection):
+        """Open a connection on the adapter.
+
+        This may mutate the given connection (in particular, its state and its
+        handle).
+
+        This should be thread-safe, or hold the lock if necessary. The given
+        connection should not be in either in_use or available.
+
+        :param Connection connection: A connection object to open.
+        :return: A connection with a handle attached and an 'open' state.
+        :rtype: Connection
+        """
+        raise dbt.exceptions.NotImplementedException(
+            '`open` is not implemented for this adapter!'
+        )
+
+    def assign(self, conn, name):
+        """Open a connection if it's not already open, and assign it name
+        regardless.
+
+        The caller is responsible for putting the assigned connection into the
+        in_use collection.
+
+        :param Connection conn: A connection, in any state.
+        :param str name: The name of the connection to set.
+        """
+        if name is None:
+            name = 'master'
+
+        conn.name = name
+
+        if conn.state == 'open':
+            logger.debug('Re-using an available connection from the pool.')
+        else:
+            logger.debug('Opening a new connection, currently in state {}'
+                         .format(conn.state))
+            conn = self.open(conn)
+
+        return conn
+
+    def release(self, name):
+        with self.lock:
+            if name not in self.in_use:
+                return
+
+            to_release = self.in_use.pop(name)
+            # to_release is temporarily neither in use nor available, but both
+            # collections are in a sane state, so we can release the lock.
```

Comment:
Good point, it would! I think any failure would be fatal, but I'll add handling just in case.

I think it'll be something like "if rollback/close throws an exception, build a new connection and add it back in instead".